AP_Notify: Support for OLED display by Alexey Kozin (attempt 3)

rmackay9 opened the pull request:
This is a replacement for these earlier PRs:
https://github.com/ArduPilot/ardupilot/pull/5135
https://github.com/ArduPilot/ardupilot/pull/5082

This PR adds:

- driver for the SH1106 OLED Display (previously only SSD1306 was supported)
- new NTF_DISPLAY_TYPE parameter allows selecting the Display type (0=disabled which is also the default)
- allows using the display on the Pixhawk flight controller

The estimated I2C traffic is 16k bytes / second while the vehicle is disarmed.  The display is not updated while the vehicle is armed except for a one time transfer (of 8k bytes) within 0.5sec of arming to display ">>> ARMED <<<" on the display.
If the I2C bus is running at the lower 100kbits/sec I think this means it consumes more than the entire bandwidth of the I2C bus.  If it's running at 400kbits/sec I think it consumes about 1/3rd of the entire bandwidth.  Is my math wrong perhaps?

The display shows:

- pre-arm failure messages
- flight mode
- GPS Hdop and number of satellites
- Prearm: passed/failed
- EKF: ok/bad

Here is a known working display: https://www.aliexpress.com/item/Wholesale-0-96-inch-4pin-White-OLED-Module-SSD1306-Drive-IC-128-64-I2C-IIC-Communication/32658908775.html
The Adafruit displays do not work because they use 0x3D as their I2C address.  They seem to display random dots as well for unknown reasons: https://www.adafruit.com/product/326

rmackay9 commented:
A couple of extra comments:

- the Adafruit modules use I2C address 0x3D (one higher than the default)
- when using the Adafruit module I get random dots as reported by @mariansoban.

rmackay9 commented:
Re the time the I2C transfers take, I've tried plugging in an external RGB LED, external compass and one of these screens to a pixhawk (using an I2C splitter) and haven't noticed any complaints or weird behaviour from any of them.  I will ask @tridge tomorrow if there is a better way to test the I2C bandwidth usage.

dipspb commented:
Why do we need https://github.com/ArduPilot/ardupilot/pull/5135 get closed and new PR created for same functionality?

lucasdemarchi commented:
> Why do we need #5135 get closed and new PR created for same functionality?


@dipspb because we cannot push to your branch without creating a new PR... @rmackay9 added some commits on top in order to expedite the merge of this functionality. You are still the author of the commits you sent.

dipspb commented:
@lucasdemarchi Thank you, now I can see it. About authorship it isn't a big problem but it seems not exactly the case. Contribution of SH1106 support was mine. Anyway it isn't a problem at all.

lucasdemarchi commented on ArduCopter/sensors.cpp:
```diff
@@ -166,6 +166,7 @@ void Copter::read_battery(void)
     // update motors with voltage and current
     if (battery.get_type() != AP_BattMonitor::BattMonitor_TYPE_NONE) {
         motors->set_voltage(battery.voltage());
+        AP_Notify::flags.battery_voltage = battery.voltage(); //for on-board oled display
```
could we get rid of all these "//for on-board oled display" comments? It's very clear what's the purpose now and later on this comment will become outdated when we introduce e.g. text-to-speech.

gmorph replied:
I like the comments - makes it much clearer.  I'm not sure how its obvious that 
AP_Notify::flags.battery_voltage = battery.voltage();
is for the oled only but perhaps I'm missing something.

lucasdemarchi replied:
it may be for oled... it may be for something else. As far as this called is concerned, it's passing the info to the "notify" object. Notify object could have a buzzer, a oled display, a monitor, a text-to-speech interface or anything else. All of these "anything else" is not interesting to this caller.

It's very frequent that we later introduce more "backends"  and these comments just add noise to the codebase as they get outdated.

gmorph replied:
Thanks for clarifying Lucas.

rmackay9 replied:
Ok, i've removed these comments in two follow up commits.

lucasdemarchi commented on libraries/AP_Notify/AP_Notify.cpp:
```diff
@@ -31,6 +31,8 @@
 #include "VRBoard_LED.h"
 #include "DiscreteRGBLed.h"
 #include "DiscoLED.h"
+#include <stdio.h>  // for strncpy
```
and also this one.

lucasdemarchi commented on libraries/AP_Notify/AP_Notify.h:
```diff
@@ -111,10 +121,17 @@ class AP_Notify
 private:
     static NotifyDevice* _devices[];
 
+    static char _send_text[NOTIFY_TEXT_BUFFER_SIZE];
```
why is this forced to be static?  The entire object is already static.

rmackay9 replied:
I've made this _send_text a non-static string in a follow-up commit.

lucasdemarchi commented on libraries/AP_Notify/Display_SH1106_I2C.cpp:
```diff
@@ -0,0 +1,136 @@
+
+    if (success) {
+        _need_hw_update = true;
+        _dev->register_periodic_callback(20000, FUNCTOR_BIND_MEMBER(&Display_SH1106_I2C::_timer, void));
```
why do we do 50Hz here if the hw_update() from the main thread is at 2Hz?

dipspb replied:
IMHO it is 50HZ because we need to refresh a first line of display with ticker running there and showing messages that doesn't fin into visible display width.

dipspb replied:
No, I was wrong. It is not the case.

rmackay9 replied:
I think it's like this to reduce lag between when the display_buffer is updated and when it appears on the screen.  The way it is now, there is at most 1/50th of a second between the display-buffer being updated and when the screen starts updating.  If we reduced the timer to 2hz then the lag could climb as high as 0.25sec.

lucasdemarchi commented on libraries/AP_Notify/Display_SH1106_I2C.h:
```diff
@@ -0,0 +1,26 @@
+
+    AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev;
+    uint8_t _displaybuffer[SH1106_COLUMNS * SH1106_ROWS_PER_PAGE];
+    bool _need_hw_update;
```
technically this should be `std::atomic<bool> _need_hw_update`, but not that important for this type of device since it will get the update next time the function is called.

lucasdemarchi commented on libraries/AP_Notify/Display_SH1106_I2C.cpp:
```diff
@@ -0,0 +1,136 @@
+    struct PACKED {
+        uint8_t reg;
+        uint8_t cmd[3];
+    } command = { 0x0, {0x02 /* |= column[0:3] */, 0x10 /* |= column[4:7] */, 0xB0 /* |= page */} };
```
idea:
```c++
struct PACKED {
    uint8_t reg;
    uint8_t column_0_3;
    uint8_t column_4_7;
    uint8_t page;
} command = { 0x0, 0x2, 0x10, 0xB0 };
```

this way you don't need comments since the variable name already documents what it is.

rmackay9 replied:
changed the command structure, thanks.

lucasdemarchi commented on libraries/AP_Notify/Display_SH1106_I2C.cpp:
```diff
@@ -0,0 +1,136 @@
+        command.cmd[2] = 0xB0 | (i & 0x0F);
+        _dev->transfer((uint8_t *)&command, sizeof(command), nullptr, 0);
+
+        memcpy(&display_buffer.db[0], &_displaybuffer[i * SH1106_COLUMNS], SH1106_COLUMNS/2);
```
here you are accessing _displaybuffer that is also used by the main thread. You need a lock to protect those accesses. In the beggining of this function you could do:
```c++
_mutex.lock();
memcpy(displaybuffer, _displaybuffer, sizeof(_displaybuffer));
_mutex.unlock();
```

or something like this.

rmackay9 replied:
I'm not familiar with how the _mutex should be initialised so I failed when I tried to add this.  Any advice welcome.

lucasdemarchi commented:
I added some comments above.

dipspb commented on libraries/AP_Notify/Display.cpp:
```diff
@@ -279,26 +282,122 @@ static const uint8_t _font[] = {
+    "GUID",
+    "INIT"
+};
+#endif
```
There is a difference between same names used in vehicle and display. Names used in display are kind of abbreviated due to limited space for name on display. So there will be needed two functions instead:
*AP_Vehicle::get_flight_mode_str()
and
*AP_Vehicle::get_flight_mode_short_str()

rmackay9 commented on libraries/AP_Notify/Display.cpp:
```diff
@@ -279,26 +282,122 @@ static const uint8_t _font[] = {
+            break;
+    }
+
+    if (_driver != nullptr) {
```
ok, changed in a follow-up commit.

rmackay9 commented on libraries/AP_Notify/Display.cpp:
```diff
@@ -279,26 +282,122 @@ static const uint8_t _font[] = {
 void Display::update()
 {
     static uint8_t timer = 0;
+    static uint8_t screenpage =0;
```
ok, moved to be part of the class in a follow up commit.

rmackay9 commented on libraries/AP_Notify/Display_Backend.h:
```diff
@@ -0,0 +1,19 @@
+#define OLED_I2C_BUS 2
+#else
+#define OLED_I2C_BUS 1
+#endif
```
ok, follow-up commit added in which the bus and address are used in the frontend and the device is pass into the constructor.

rmackay9 commented on libraries/AP_Notify/Display_SH1106_I2C.cpp:
```diff
@@ -0,0 +1,136 @@
+
+bool Display_SH1106_I2C::hw_init()
+{
+    struct {
```
thanks, added the PACKED.

rmackay9 commented:
@lucasdemarchi, I think I've addressed all the issues you've raised with the exception of the _mutex thing.  Are you happy for this to go in and/or can you educate me on how the mutex thing can be used?

rmackay9 commented:
@lucasdemarchi, I've added the semaphore now as well..

rmackay9 commented:
Thanks @lucasdemarchi !!

dipspb commented:
Thank you all very much for your great support and assistance with this feature!
